File: runtime/pluto/src/main/java/mb/pipe/run/pluto/spoofax/Parse.java
package mb.pipe.run.pluto.spoofax;

import java.io.File;
import java.io.IOException;

import javax.annotation.Nullable;

import org.apache.commons.vfs2.FileObject;
import org.metaborg.core.language.ILanguageImpl;
import org.metaborg.core.language.LanguageIdentifier;
import org.metaborg.spoofax.core.syntax.SyntaxFacet;
import org.metaborg.spoofax.core.unit.ISpoofaxInputUnit;
import org.metaborg.spoofax.core.unit.ISpoofaxParseUnit;
import org.spoofax.interpreter.terms.IStrategoTerm;

import build.pluto.builder.BuildRequest;
import build.pluto.builder.Builder;
import build.pluto.builder.factory.BuilderFactory;
import build.pluto.dependency.Origin;
import mb.pipe.run.core.model.IContext;
import mb.pipe.run.core.vfs.IResource;
import mb.pipe.run.pluto.util.ABuilder;
import mb.pipe.run.pluto.util.AInput;
import mb.pipe.run.pluto.util.Result;

public class Parse extends ABuilder<Parse.Input, Parse.Output> {
    public static class Input extends AInput {
        private static final long serialVersionUID = 1L;

        public final LanguageIdentifier langId;
        public final IResource file;
        public final String text;


        public Input(IContext context, @Nullable Origin origin, LanguageIdentifier langId, IResource file, String text) {
            super(context, origin);

            this.langId = langId;
            this.file = file;
            this.text = text;
        }
    }

    public static class Output implements build.pluto.output.Output {
        private static final long serialVersionUID = 1L;

        public final @Nullable IStrategoTerm ast;


        public Output(@Nullable IStrategoTerm ast) {
            this.ast = ast;
        }


        @Override public int hashCode() {
            final int prime = 31;
            int result = 1;
            result = prime * result + ((ast == null) ? 0 : ast.hashCode());
            return result;
        }

        @Override public boolean equals(Object obj) {
            if(this == obj)
                return true;
            if(obj == null)
                return false;
            if(getClass() != obj.getClass())
                return false;
            Output other = (Output) obj;
            if(ast == null) {
                if(other.ast != null)
                    return false;
            } else if(!ast.equals(other.ast))
                return false;
            return true;
        }
    }


    public static final BuilderFactory<Input, Output, Parse> factory = factory(Parse.class, Input.class);

    public static BuildRequest<Input, Output, Parse, BuilderFactory<Input, Output, Parse>> request(Input input) {
        return request(input, Parse.class, Input.class);
    }

    public static Origin origin(Input input) {
        return origin(input, Parse.class, Input.class);
    }

    public static Result<Output> requireBuild(Builder<?, ?> requiree, Input input) throws IOException {
        return requireBuild(requiree, input, Parse.class, Input.class);
    }

    public static @Nullable IStrategoTerm build(Builder<?, ?> requiree, Input input) throws IOException {
        return requireBuild(requiree, input, Parse.class, Input.class).output.ast;
    }


    public Parse(Input input) {
        super(input);
    }


    @Override protected String description(Input input) {
        return "Parse text of " + input.langId;
    }

    @Override public File persistentPath(Input input) {
        return depFile("parse", input.langId, input.file, input.text);
    }

    @Override protected Output build(Input input) throws Throwable {
        requireOrigins();

        final ILanguageImpl langImpl = spoofax().languageService.getImpl(input.langId);
        final SyntaxFacet facet = langImpl.facet(SyntaxFacet.class);
        if(facet != null) {
            final FileObject parseTableFile = facet.parseTable;
            if(parseTableFile != null) {
                final File localFile = spoofax().resourceService.localPath(parseTableFile);
                if(localFile != null) {
                    require(localFile);
                }
            }
        }
        final FileObject resource = input.file.fileObject();
        final ISpoofaxInputUnit inputUnit = spoofax().unitService.inputUnit(resource, input.text, langImpl, null);
        final ISpoofaxParseUnit parseUnit = spoofax().syntaxService.parse(inputUnit);
        return new Output(parseUnit.ast());
    }
}

File: runtime/ceres/src/main/kotlin/mb/pipe/run/ceres/spoofax/core/Parse.kt
package mb.pipe.run.ceres.spoofax.core

import com.google.inject.Inject
import mb.ceres.BuildContext
import mb.ceres.Builder
import mb.ceres.PathStampers
import mb.log.Logger
import mb.pipe.run.ceres.util.Tuple3
import mb.pipe.run.core.model.message.Msg
import mb.pipe.run.core.model.parse.Token
import mb.pipe.run.spoofax.cfg.SpxCoreConfig
import mb.pipe.run.spoofax.sdf.TokenExtractor
import mb.pipe.run.spoofax.util.MessageConverter
import org.metaborg.core.syntax.ParseException
import org.metaborg.spoofax.core.syntax.SyntaxFacet
import org.spoofax.interpreter.terms.IStrategoTerm
import java.io.Serializable

class CoreParse @Inject constructor(log: Logger, val messageConverter: MessageConverter) : Builder<CoreParse.Input, CoreParse.Output> {
  companion object {
    val id = "coreParse"
  }

  data class Input(val config: SpxCoreConfig, val text: String) : Serializable
  data class Output(val ast: IStrategoTerm?, val tokens: ArrayList<Token>?, val messages: ArrayList<Msg>) : Tuple3<IStrategoTerm?, ArrayList<Token>?, ArrayList<Msg>>

  val log: Logger = log.forContext(CoreTrans::class.java)

  override val id = Companion.id
  override fun BuildContext.build(input: Input): Output {
    val spoofax = Spx.spoofax()
    val langImpl = buildOrLoad(input.config)

    // Require parse table
    val facet = langImpl.facet<SyntaxFacet>(SyntaxFacet::class.java)
    if (facet != null) {
      val parseTableFile = facet.parseTable
      if (parseTableFile != null) {
        require(parseTableFile.pPath, PathStampers.hash)
      }
    }

    // Perform parsing
    val inputUnit = spoofax.unitService.inputUnit(input.text, langImpl, null)
    try {
      val parseUnit = spoofax.syntaxService.parse(inputUnit)
      val ast = parseUnit.ast();
      val tokens = if (ast != null) TokenExtractor.extract(ast) else null
      val messages = messageConverter.toMsgs(parseUnit.messages());
      return Output(ast, tokens, messages);
    } catch(e: ParseException) {
      log.error("Parsing failed unexpectedly", e)
      return Output(null, null, ArrayList(0))
    }
  }
}

fun BuildContext.parse(input: CoreParse.Input) = requireOutput(CoreParse::class.java, input)
fun BuildContext.parse(config: SpxCoreConfig, text: String) = parse(CoreParse.Input(config, text))
